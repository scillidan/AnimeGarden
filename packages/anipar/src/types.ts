export interface ParseOptions {}

export interface ParseResult {
  title: string;

  // Other titles like translations
  titles?: string[];

  // fansub info
  fansub?: {
    // fansub name
    name: string;

    // collab fansubs name
    collab?: string[];
  };

  // season info
  season?: {
    // season number
    number: number;

    // season title
    title?: string;
  };

  // Type, e.g. OVA
  type?: string;

  // episode info
  episode?: EpisodeInfo;

  // Multiple episodes
  episodes?: EpisodeInfo[];

  // episode range
  episodeRange?: {
    from: number;

    to: number;
  };

  // Release version
  version?: string;

  // Video language
  language?: string;

  // Subtitles including language, e.g. 简日内嵌
  subtitles?: string;

  // Source, e.g. WEB-DL
  source?: string;

  // Source platform, e.g. Baha
  platform?: string;

  // Onair year
  year?: number;

  // Onair month
  month?: number;

  // Video file info
  file?: FileInfo;

  // Other tags
  tags: string[];
}

export interface EpisodeInfo {
  number: number;

  // 总集篇 xx.5
  numberSub?: number;

  // Episode title
  title?: string;
}

export interface FileInfo {
  extension?: string;

  audio?: {
    term?: string;
  };

  video?: {
    term?: string;

    resolution?: string;
  };
}
