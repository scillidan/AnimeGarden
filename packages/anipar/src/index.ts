import type { ParseOptions, ParseResult } from './types.ts';

export type { ParseResult };

export function parse(title: string, options: ParseOptions = {}): ParseResult | undefined {
  return undefined;
}
