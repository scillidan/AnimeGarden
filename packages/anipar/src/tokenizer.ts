export class Token {
  public readonly text: string;

  public readonly left: string | undefined;

  public readonly right: string | undefined;

  public constructor(text: string, left?: string, right?: string) {
    this.text = text;
    this.left = left;
    this.right = right;
  }

  public get isWrapped() {
    return this.left && this.right;
  }

  public slice(start: number, end?: number) {
    const text = this.text.slice(start, end);
    return new Token(text, this.left, this.right);
  }

  public trim() {
    const text = this.text.trim();
    return new Token(text, this.left, this.right);
  }

  public toString() {
    return `${this.left ?? ''}${this.text}${this.right ?? ''}`;
  }
}

export const Wrappers = new Map([
  ['[', ']'],
  ['【', '】'],
  ['(', ')'],
  ['（', '）'],
  ['{', '}']
]);

export const RevWrappers = new Map([...Wrappers.entries()].map(([k, v]) => [v, k]));

export function tokenize(text: string) {
  const tokens: Token[] = [];

  let cursor = 0;
  let cur = '';
  let left: string | undefined = undefined;
  let right: string | undefined = undefined;
  while (cursor < text.length) {
    const char = text[cursor];
    if (Wrappers.has(char)) {
      if (cur) {
        tokens.push(new Token(cur.trim()));
        cur = '';
      }
      left = char;
      right = Wrappers.get(char)!;
    } else if (left && right && char === right) {
      tokens.push(new Token(cur.trim(), left, right));
      cur = '';
      left = undefined;
      right = undefined;
    } else {
      cur += char;
    }
    cursor += 1;
  }
  if (cur) {
    tokens.push(new Token(cur.trim()));
  }

  return tokens.filter((t) => t.text);
}
