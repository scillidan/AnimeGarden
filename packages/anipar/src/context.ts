import type { ParseOptions, ParseResult } from './types';

import { Token } from './tokenizer';

export class Context {
  // Left cursor for consumed tokens
  public left = 0;

  // Right cursor for consumed tokens
  public right;

  public tokens: Token[];

  public options: ParseOptions;

  public result: Partial<ParseResult>;

  public tags: string[];

  public constructor(tokens: Token[], options: ParseOptions) {
    this.tokens = tokens;
    this.options = options;
    this.right = tokens.length - 1;
    this.tags = [];
    this.result = {};
  }

  public update<K1 extends keyof ParseResult>(key: K1, value: ParseResult[K1]) {
    this.result[key] = value;
  }

  public update2<
    K1 extends keyof Required<ParseResult>,
    K2 extends keyof Required<ParseResult>[K1]
  >(key1: K1, key2: K2, value: Required<ParseResult>[K1][K2]) {
    if (!this.result[key1]) {
      // @ts-expect-error
      this.result[key1] = {};
    }
    // @ts-expect-error
    this.result[key1][key2] = value;
  }

  public update3<
    K1 extends keyof Required<ParseResult>,
    K2 extends keyof Required<ParseResult>[K1],
    K3 extends keyof Required<Required<ParseResult>[K1]>[K2]
  >(key1: K1, key2: K2, key3: K3, value: Required<Required<ParseResult>[K1]>[K2][K3]) {
    if (!this.result[key1]) {
      // @ts-expect-error
      this.result[key1] = {};
    }
    // @ts-expect-error
    if (!this.result[key1][key2]) {
      // @ts-expect-error
      this.result[key1][key2] = {};
    }
    // @ts-expect-error
    this.result[key1][key2][key3] = value;
  }

  public get hasEpisode() {
    return this.result.episode || this.result.episodes || this.result.episodeRange;
  }

  public validate(): ParseResult | undefined {
    this.result.tags = this.tags;
    return this.result as ParseResult;
  }
}
