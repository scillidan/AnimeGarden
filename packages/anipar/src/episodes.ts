import { Context } from './context';

const WrappedEpisodeRE = /^(?<ep1>\d+)(?:\.(\d))?(?:[vV](\d+))?$|^第(?<ep2>\d+)[集话話]$/;
const WrappedMovieRE = /^Movie [vV](\d+)$/;

const EpisodesRange1 = /^(\d+)-(\d+)(?:\s*.*)$/;
const EpisodesRange2 = /^全(\d+)集$/;

export function parseWrappedEpisodes(ctx: Context) {
  const token = ctx.tokens[ctx.right];
  const text = token.text;
  if (token.isWrapped && matchEpiodes(ctx, text)) {
    ctx.right -= 1;
    return true;
  }
  return false;
}

export function matchEpiodes(ctx: Context, text: string) {
  // 1. Single episode
  {
    const res = WrappedEpisodeRE.exec(text);
    if (res) {
      const epText = res.groups?.ep1! || res.groups?.ep2!;
      const ep = +epText;
      if (!Number.isNaN(ep)) {
        // Handle year: [2024]
        if (1949 <= ep && ep <= 2099 && text === epText && ctx.hasEpisode) {
          ctx.update('year', ep);
          return true;
        }

        ctx.update2('episode', 'number', ep);
        // 1.5
        if (res[2] && !Number.isNaN(res[2])) {
          ctx.update2('episode', 'numberSub', +res[2]);
        }
        // 3v2
        if (res[3] && !Number.isNaN(res[3])) {
          ctx.update('version', +res[3]);
        }
        return true;
      }
    }
  }
  {
    // [Movie] / [Movie v2]
    const res = WrappedMovieRE.exec(text);
    if (res) {
      if (res[1] && !Number.isNaN(res[1])) {
        ctx.update('version', +res[1]);
      }
      ctx.update('type', 'Movie');
      return true;
    }
  }

  // 2. Episode range
  {
    // 01-26
    const res = EpisodesRange1.exec(text);
    if (res) {
      const from = +res[1];
      const to = +res[2];
      if (!Number.isNaN(from) && !Number.isNaN(to)) {
        ctx.update2('episodeRange', 'from', from);
        ctx.update2('episodeRange', 'to', to);
        return true;
      }
    }
  }
  {
    // 全26集
    const res = EpisodesRange2.exec(text);
    if (res) {
      const to = +res[1];
      if (!Number.isNaN(to)) {
        ctx.update2('episodeRange', 'from', 1);
        ctx.update2('episodeRange', 'to', to);
        return true;
      }
    }
  }

  return false;
}

const SuffixEpisodeRE = [/- (\d+)(?:\.(\d))?$/, /第(\d+)(?:\.(\d))?[集话話]$/];

export function parseSuffixEpisodes(ctx: Context) {
  if (ctx.result.episode || ctx.result.episodes || ctx.result.episodeRange) {
    return true;
  }

  const token = ctx.tokens[ctx.right];
  const text = token.text;
  for (const RE of SuffixEpisodeRE) {
    const res = RE.exec(text);
    if (res) {
      const ep = +res[1];
      if (!Number.isNaN(ep)) {
        ctx.update2('episode', 'number', ep);

        // 1.5
        if (res[2] && !Number.isNaN(res[2])) {
          ctx.update2('episode', 'numberSub', +res[2]);
        }

        // Remove token suffix
        ctx.tokens[ctx.right] = token.slice(0, text.length - res[0].length).trim();

        return true;
      }
    }
  }
}

// Season regexp
export const SuffixSeasonOrEpisodesRes: Array<
  [RegExp, (res: RegExpExecArray, ctx: Context) => boolean]
> = [
  [
    /(?:S|Season\s?)(\d+)$/,
    (res, ctx) => {
      const season = +res[1];
      if (!Number.isNaN(season)) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /(1st|2nd|3rd|[456789]th) Season$/,
    (res, ctx) => {
      const season = Number.parseInt(res[1]);
      if (!Number.isNaN(season)) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /第?(\d+)[季期]$/,
    (res, ctx) => {
      const season = +res[1];
      if (!Number.isNaN(season)) {
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /第?((?:[零一二三四五六七八九]十)?[零一二三四五六七八九])[季期]$/,
    (res, ctx) => {
      const text = res[1];
      const map = {
        零: 0,
        一: 1,
        二: 2,
        三: 3,
        四: 4,
        五: 5,
        六: 6,
        七: 7,
        八: 8,
        九: 9
      };
      const base =
        text.length === 2 && text[0] === '十'
          ? 1
          : text.length === 3
            ? map[text[0] as keyof typeof map]!
            : 0;
      const offset = map[text[text.length - 1] as keyof typeof map];
      const season = base * 10 + offset;
      if (!ctx.result.season?.number || ctx.result.season.number === season) {
        ctx.update2('season', 'number', season);
        return true;
      }
      return false;
    }
  ],
  [
    / (\d+)$/,
    (res, ctx) => {
      const season = +res[1];
      if (
        !Number.isNaN(season) &&
        (ctx.result.episode || ctx.result.episodes || ctx.result.episodeRange)
      ) {
        // Skip: Part 2
        if (/Parts? \d+$/.test(res.input)) {
          return false;
        }
        if (!ctx.result.season?.number || ctx.result.season.number === season) {
          ctx.update2('season', 'number', season);
          return true;
        }
      }
      return false;
    }
  ],
  [
    /(?<ep1>\d+)(?:\.(\d))?(?:[vV](\d+))?$|^第(?<ep2>\d+)[集话話]$/,
    (res, ctx) => {
      const ep = +(res.groups?.ep1! || res.groups?.ep2!);
      if (!Number.isNaN(ep)) {
        // Skip: Part 2
        if (/Parts? \d+$/.test(res.input)) {
          return false;
        }

        ctx.update2('episode', 'number', ep);
        // 1.5
        if (res[2] && !Number.isNaN(res[2])) {
          ctx.update2('episode', 'numberSub', +res[2]);
        }
        // 3v2
        if (res[3] && !Number.isNaN(res[3])) {
          ctx.update('version', +res[3]);
        }
        return true;
      }
      return false;
    }
  ],
  [
    /（日配版）$/,
    (res, ctx) => {
      ctx.update('language', res[0]);
      return true;
    }
  ],
  [
    /（重播）$/,
    (res, ctx) => {
      ctx.tags.push(res[0]);
      return true;
    }
  ]
];
