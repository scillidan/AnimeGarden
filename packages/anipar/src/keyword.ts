import { Context } from './context';

const AudioTerm = new Set([
  // Audio channels
  '2.0CH',
  '2CH',
  '5.1',
  '5.1CH',
  'DTS',
  'DTS-ES',
  'DTS5.1',
  'TRUEHD5.1',
  // Audio codec
  'AAC',
  'AACX2',
  'AAC×2',
  'AACX3',
  'AAC×3',
  'AACX4',
  'AAC×4',
  'AC3',
  'EAC3',
  'E-AC-3',
  'FLAC',
  'FLACX2',
  'FLAC×2',
  'FLACX3',
  'FLAC×3',
  'FLACX4',
  'FLAC×4',
  'LOSSLESS',
  'MP3',
  'OGG',
  'VORBIS',
  // Audio language
  'DUALAUDIO',
  'DUAL AUDIO'
]);

const VideoTerm = new Set([
  // Frame rate
  '23.976FPS',
  '24FPS',
  '29.97FPS',
  '30FPS',
  '60FPS',
  '120FPS',
  // Video codec
  '8BIT',
  '8-BIT',
  '10BIT',
  '10BITS',
  '10-BIT',
  '10-BITS',
  'HI10',
  'HI10P',
  'HI444',
  'HI444P',
  'HI444PP',
  'H264',
  'H265',
  'H.264',
  'H.265',
  'X264',
  'X265',
  'X.264',
  'AVC',
  'HEVC',
  'HEVC2',
  'HEVC-10BIT',
  'DIVX',
  'DIVX5',
  'DIVX6',
  'XVID',
  // Video format
  'AVI',
  'RMVB',
  'WMV',
  'WMV3',
  'WMV9',
  // Video quality
  'HQ',
  'LQ',
  // Video resolution
  'HD',
  'SD'
]);

const VideoResolution = new Set([
  '480P',
  '720P',
  '1080P',
  '2160P',
  '1280X720',
  '1280×720',
  '1920X1080',
  '1920×1080',
  '2K',
  '4K'
]);

const Source = new Set([
  'BD',
  'BDRIP',
  'BLURAY',
  'BLU-RAY',
  'DVD',
  'DVD5',
  'DVD9',
  'DVD-R2J',
  'DVDRIP',
  'DVD-RIP',
  'R2DVD',
  'R2J',
  'R2JDVD',
  'R2JDVDRIP',
  'HDTV',
  'HDTVRIP',
  'TVRIP',
  'TV-RIP',
  'WEBCAST',
  'WEBDL',
  'WEB-DL',
  'WEBRIP',
  'WEB-RIP',
  'WEB-MKV'
]);

const Platfrom = new Set(['Baha', 'B-Global', 'ABEMA', 'CR', 'ViuTV', 'AMZN', 'ADN']);

const Type = new Set([
  'GEKIJOUBAN',
  'MOVIE',
  'OAD',
  'OAV',
  'ONA',
  'OVA',
  'SPECIAL',
  'SPECIALS',
  'TV',
  '特别篇',
  '特別篇',
  '特別編',
  '特别话',
  '特別话',
  '特別話',
  '番外篇',
  '番外編',
  '剧场版',
  '劇場版',
  //
  'SP',
  //
  'ED',
  'ENDING',
  'NCED',
  'NCOP',
  'OP',
  'OPENING',
  'PREVIEW',
  'PV'
]);

const Languages = new Set(['CHS', 'CHT', '简体', '国语中字', '繁體', '简日双语', '繁日雙語']);

const Subtitles = new Set([
  'ASS',
  'GB',
  'BIG5',
  'DUB',
  'DUBBED',
  'HARDSUB',
  'HARDSUBS',
  'RAW',
  'SOFTSUB',
  'SOFTSUBS',
  'SUB',
  'SUBBED',
  'SUBTITLED'
]);

const LanguagePrefixes = [
  '简体',
  '繁體',
  '简日',
  '繁日',
  '简繁日双语',
  '简繁日',
  '简繁',
  '繁简日',
  '简繁日语',
  '简日双语'
];

const SubtitlesSufixes = new Set(['内嵌', '內嵌', '内封', '内封字幕', '外挂', '外掛']);

const LanguageSubtitles = new Map([
  ['粵日雙語+內封繁體中文字幕', ['繁體中文', '內封字幕']],
  ['粵語+無對白字幕', [undefined, '無對白字幕']]
]);

const Extension = new Set([
  '3GP',
  'AVI',
  'DIVX',
  'FLV',
  'M2TS',
  'MKV',
  'MOV',
  'MP4',
  'MPG',
  'OGM',
  'RM',
  'RMVB',
  'TS',
  'WEBM',
  'WMV'
]);

// Prefix
const SearchPrefix = ['检索：', '检索用：'];
const HiringPrefix = ['招募', '字幕社招人'];

function matchSingleTag(ctx: Context, text: string) {
  const upper = text.toUpperCase();
  // Match keywords
  if (AudioTerm.has(upper)) {
    ctx.update3('file', 'audio', 'term', text);
    return true;
  }
  if (VideoTerm.has(upper)) {
    ctx.update3('file', 'video', 'term', text);
    return true;
  }
  if (VideoResolution.has(upper)) {
    ctx.update3('file', 'video', 'resolution', text);
    return true;
  }
  if (Source.has(upper)) {
    ctx.update('source', text);
    return true;
  }
  if (Platfrom.has(text)) {
    ctx.update('platform', text);
    return true;
  }
  if (Type.has(upper)) {
    ctx.update('type', text);
    return true;
  }
  if (Extension.has(upper)) {
    ctx.update2('file', 'extension', text);
    return true;
  }

  // Match language and subtitles
  {
    if (Languages.has(upper)) {
      ctx.update('language', text);
      return true;
    }
    if (Subtitles.has(upper)) {
      ctx.update('subtitles', text);
      return true;
    }
    // Combine language and subtitles
    for (const prefix of LanguagePrefixes) {
      if (text.startsWith(prefix)) {
        const language = prefix;
        const subtitles = text.slice(prefix.length);
        if (SubtitlesSufixes.has(subtitles)) {
          ctx.update('language', language);
          ctx.update('subtitles', subtitles);
          return true;
        }
      }
    }
    const combined = LanguageSubtitles.get(text);
    if (combined) {
      ctx.update('language', combined[0]);
      ctx.update('subtitles', combined[1]);
      return true;
    }
  }

  // Match regex
  {
    {
      // 2024年10月番
      const match = /^(\d\d\d\d)年(\d\d?)月新?番$/.exec(text);
      if (match) {
        const year = +match[1];
        const month = +match[2];
        ctx.update('year', year);
        ctx.update('month', month);
        return true;
      }
    }
    {
      // ★10月新番
      const match = /^★?(\d\d?)月新?番★?$/.exec(text);
      if (match) {
        const month = +match[1];
        ctx.update('month', month);
        return true;
      }
    }
    {
      // v2
      const match = /^[vV](\d+)$/.exec(text);
      if (match) {
        const version = +match[1];
        ctx.update('version', version);
        return true;
      }
    }
  }

  // Match prefix
  {
    for (const prefix of SearchPrefix) {
      if (text.startsWith(prefix)) {
        const title = text.slice(prefix.length).trim();
        ctx.tags.push(title);
        return true;
      }
    }
    for (const prefix of HiringPrefix) {
      if (text.startsWith(prefix)) {
        return true;
      }
    }
  }

  return false;
}

const TagSeperators = [' ', '_'];
function matchMultipleTags(ctx: Context, text: string) {
  for (const sep of TagSeperators) {
    const parts = text.split(sep);
    if (parts.length <= 1) continue;
    let matched = true;
    for (let i = 0; i < parts.length; i += 1) {
      const part = parts[i];
      if (!matchSingleTag(ctx, part)) {
        matched = false;
      }
    }
    if (matched) {
      return true;
    }
  }
  return false;
}

function parseTag(ctx: Context, cursor: number) {
  const token = ctx.tokens[cursor];
  if (token.isWrapped) {
    const text = token.text;
    if (matchSingleTag(ctx, text)) {
      return true;
    }
    if (matchMultipleTags(ctx, text)) {
      return true;
    }
  }
  return false;
}

export function parseRightTags(ctx: Context) {
  while (ctx.right > ctx.left) {
    if (parseTag(ctx, ctx.right)) {
      ctx.right -= 1;
    } else {
      break;
    }
  }
}

export function parseLeftTags(ctx: Context) {
  while (ctx.left < ctx.right) {
    if (parseTag(ctx, ctx.left)) {
      ctx.left += 1;
    } else {
      break;
    }
  }

  // Match prefix
  {
    const token = ctx.tokens[ctx.left];
    const text = token.text;
    // 【极影字幕社】 ★10月新番
    const match = /^★?(\d\d?)月新?番★?/.exec(text);
    if (match) {
      const matched = match[0];
      const month = +match[1];
      ctx.update('month', month);
      ctx.tokens[ctx.left] = token.slice(matched.length);
    }
  }
  {
    const token = ctx.tokens[ctx.left];
    const text = token.text;
    const match = /^★?(剧场版|劇場版)★?/.exec(text);
    if (match) {
      const matched = match[0];
      const type = match[1];
      ctx.update('type', type);
      ctx.tokens[ctx.left] = token.slice(matched.length);
    }
  }
}
