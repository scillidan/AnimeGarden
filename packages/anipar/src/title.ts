import { Context } from './context';
import { parseSuffixEpisodes, parseSuffixSeasonOrEpisodes } from './episodes';

export function parseFansub(ctx: Context) {
  // [fansub] title
  if (ctx.left + 1 > ctx.right) {
    return false;
  }

  const token = ctx.tokens[ctx.left];
  if (token.isWrapped) {
    const text = token.text;

    let found = false;
    const seps = ['&', '·'];
    for (const sep of seps) {
      const [name, ...collab] = text.split(sep);

      if (collab.length > 0) {
        // Hack: [jibaketa合成&壓制]
        if (collab.length === 1 && collab[0] === '壓制') {
          break;
        }

        found = true;
        ctx.update2('fansub', 'name', name);
        ctx.update2('fansub', 'collab', collab);
        break;
      }
    }

    if (!found) {
      ctx.update2('fansub', 'name', text);
    }

    ctx.left += 1;

    return true;
  }

  return false;
}

export function parseTitle(ctx: Context) {
  // 1. Parse suffix episodes
  parseSuffixEpisodes(ctx);

  // 2. Parse titles
  const rest = ctx.tokens.slice(ctx.left, ctx.right + 1);
  if (rest.length === 0) return false;

  const text = rest.length === 1 ? rest[0].text : rest.map((t) => t.toString()).join('');

  // 2.1. Try split multiple titles
  let found = false;
  const separators = ['/', '\\'];
  for (const sep of separators) {
    const parts = text
      .split(sep)
      .map((t) => t.trim())
      .filter((t) => !!t);
    if (parts.length > 1) {
      const [title, ...other] = parts;

      const trimmedTitle = parseSuffixSeasonOrEpisodes(ctx, title);
      const trimmedOther = other
        .map((t) => parseSuffixSeasonOrEpisodes(ctx, t))
        .map((t) => t.trim())
        .filter((t) => !!t);

      ctx.update('title', trimmedTitle);
      if (trimmedOther.length > 0) {
        ctx.update('titles', trimmedOther);
      }
      found = true;
    }
  }
  if (!found) {
    const trimmed = parseSuffixSeasonOrEpisodes(ctx, text);
    ctx.update('title', trimmed);
  }

  return true;
}
