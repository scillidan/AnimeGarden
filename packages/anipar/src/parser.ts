import type { ParseOptions, ParseResult } from './types';

import { Context } from './context';
import { tokenize } from './tokenizer';
import { parseLeftTags, parseRightTags } from './keyword';
import { parseFansub, parseTitle } from './title';

export function parse(title: string, options: ParseOptions = {}): ParseResult | undefined {
  const tokens = tokenize(title);
  if (tokens.length === 0) return undefined;

  const context = new Context(tokens, options);

  // 1. Parse right tags
  parseRightTags(context);
  // 2. Parse left tags
  parseLeftTags(context);
  // 3. Parse fansub
  parseFansub(context);
  parseLeftTags(context);
  // 4. Parse title
  if (!parseTitle(context)) {
    return undefined;
  }

  // 6. Return result
  const result = context.validate();
  return result ? postprocess(result) : undefined;
}

function postprocess(result: ParseResult) {
  if (result.fansub?.name === 'ANi') {
    if (result.title && result.titles && result.titles.length === 1) {
      const t1 = result.title;
      const t2 = result.titles[0];
      result.title = t2;
      result.titles = [t1];
    }
  }
  return result;
}
