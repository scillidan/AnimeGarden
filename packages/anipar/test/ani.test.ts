import { describe, it, expect } from 'vitest';

import { parse } from '../src';

const titles = [
  '[ANi] Amagamisan Chi no Enmusubi / 結緣甘神神社 - 12 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4]',
  '[ANi] 妖怪學校的菜鳥老師 - 11 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4]',
  '[ANi] Seirei Gensouki / 精靈幻想記 2 - 11 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4]',
  '[ANi] 最狂輔助職業【話術士】世界最強戰團聽我號令 - 12 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4]',
  '[ANi] 鴨乃橋論的禁忌推理 - 24 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4]'
];

describe('ANi', () => {
  for (const title of titles) {
    it(title, () => {
      expect(parse(title)).toMatchSnapshot();
    });
  }
});
