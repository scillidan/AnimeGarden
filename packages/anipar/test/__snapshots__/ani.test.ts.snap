// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ANi > [ANi] Amagamisan Chi no Enmusubi / 結緣甘神神社 - 12 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4] 1`] = `
{
  "episode": {
    "number": 12,
  },
  "fansub": {
    "name": "ANi",
  },
  "file": {
    "audio": {
      "term": "AAC",
    },
    "extension": "MP4",
    "video": {
      "resolution": "1080P",
      "term": "AVC",
    },
  },
  "language": "CHT",
  "platform": "Baha",
  "source": "WEB-DL",
  "tags": [],
  "title": "結緣甘神神社",
  "titles": [
    "Amagamisan Chi no Enmusubi",
  ],
}
`;

exports[`ANi > [ANi] Seirei Gensouki / 精靈幻想記 2 - 11 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "ANi",
  },
  "file": {
    "audio": {
      "term": "AAC",
    },
    "extension": "MP4",
    "video": {
      "resolution": "1080P",
      "term": "AVC",
    },
  },
  "language": "CHT",
  "platform": "Baha",
  "season": {
    "number": 2,
  },
  "source": "WEB-DL",
  "tags": [],
  "title": "精靈幻想記",
  "titles": [
    "Seirei Gensouki",
  ],
}
`;

exports[`ANi > [ANi] 妖怪學校的菜鳥老師 - 11 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4] 1`] = `
{
  "episode": {
    "number": 11,
  },
  "fansub": {
    "name": "ANi",
  },
  "file": {
    "audio": {
      "term": "AAC",
    },
    "extension": "MP4",
    "video": {
      "resolution": "1080P",
      "term": "AVC",
    },
  },
  "language": "CHT",
  "platform": "Baha",
  "source": "WEB-DL",
  "tags": [],
  "title": "妖怪學校的菜鳥老師",
}
`;

exports[`ANi > [ANi] 最狂輔助職業【話術士】世界最強戰團聽我號令 - 12 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4] 1`] = `
{
  "episode": {
    "number": 12,
  },
  "fansub": {
    "name": "ANi",
  },
  "file": {
    "audio": {
      "term": "AAC",
    },
    "extension": "MP4",
    "video": {
      "resolution": "1080P",
      "term": "AVC",
    },
  },
  "language": "CHT",
  "platform": "Baha",
  "source": "WEB-DL",
  "tags": [],
  "title": "最狂輔助職業【話術士】世界最強戰團聽我號令",
}
`;

exports[`ANi > [ANi] 鴨乃橋論的禁忌推理 - 24 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4] 1`] = `
{
  "episode": {
    "number": 24,
  },
  "fansub": {
    "name": "ANi",
  },
  "file": {
    "audio": {
      "term": "AAC",
    },
    "extension": "MP4",
    "video": {
      "resolution": "1080P",
      "term": "AVC",
    },
  },
  "language": "CHT",
  "platform": "Baha",
  "source": "WEB-DL",
  "tags": [],
  "title": "鴨乃橋論的禁忌推理",
}
`;
